dap/adapter.go

```diff
@@ -468,3 +468,100 @@ func (s *sourceMap) Get(fname string) ([]byte, bool) {
 	}
 	return v.([]byte), true
 }
+
+type breakpointMap struct {
+	byPath map[string][]dap.Breakpoint
+	mu     sync.RWMutex
+
+	nextID atomic.Int64
+}
+
+func newBreakpointMap() *breakpointMap {
+	return &breakpointMap{
+		byPath: make(map[string][]dap.Breakpoint),
+	}
+}
+
+func (b *breakpointMap) Set(fname string, sbps []dap.SourceBreakpoint) (breakpoints []dap.Breakpoint) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	prev := b.byPath[fname]
+	for _, sbp := range sbps {
+		index := slices.IndexFunc(prev, func(e dap.Breakpoint) bool {
+			return sbp.Line >= e.Line && sbp.Line <= e.EndLine && sbp.Column >= e.Column && sbp.Column <= e.EndColumn
+		})
+
+		var bp dap.Breakpoint
+		if index >= 0 {
+			bp = prev[index]
+		} else {
+			bp = dap.Breakpoint{
+				Id:        int(b.nextID.Add(1)),
+				Line:      sbp.Line,
+				EndLine:   sbp.Line,
+				Column:    sbp.Column,
+				EndColumn: sbp.Column,
+			}
+		}
+		breakpoints = append(breakpoints, bp)
+	}
+	b.byPath[fname] = breakpoints
+	return breakpoints
+}
+
+func (b *breakpointMap) Intersect(ctx Context, src *pb.Source, ws string) map[digest.Digest]int {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	digests := make(map[digest.Digest]int)
+
+NEXT_DIGEST:
```

Comment:
style: no caps-only

Reply:
I just refactored this to put the inner for loop into another function to avoid the need for a labeled continue.